### PR TITLE
Bug Fixes and New Features

### PR DESCRIPTION
**Pull request by matt1003**
Hey @ronakg,

I love this plugin and use it every day. 

Over the last few months I have been putting together a few big fixes and also some new features. Sorry to dump them on you all at once, but I wanted to ensure that they were tested properly before I created this pull request. I have tried to break up the changes into logical patches, and provide detailed descriptions for each. Hopefully it is not too hard to follow.

Please let me know what you think, and what areas need improvement.

Thanks.
-- Matt

### CODE REVIEW

**Comment by ronakg** [comment]
This will take a while to go through. Give me a week or two.

Thanks for all the work. Looks promising.

**Comment by matt1003** [comment]
No problem. Take all the time you need.

**Comment by ronakg** [comment]
Thanks for the pr. This is great.
